src/merkle.rs

```diff
@@ -0,0 +1,148 @@
+use sha3::{Digest, Sha3_256};
+
+type Hash = Vec<u8>;
+
+struct Merkle {
+    tree: Vec<Hash>, // This will be a binary tree represented as a vector
+}
+
+fn hash(element: &[u8]) -> Hash {
+    Sha3_256::digest(element).to_vec()
+}
+
+// Helper function to hash two internal nodes
+fn hash_internal_node(left: &Hash, right: &Hash) -> Hash {
+    let mut hasher = Sha3_256::new();
+    hasher.update(left);
+    hasher.update(right);
+    hasher.finalize().to_vec()
+}
+
+impl Merkle {
+    /// Creates a new Merkle tree from a list of data elements.
+    /// TODO: Make this function generic to accept any type that can be hashed.
+    ///
+    pub fn new(data: &[Vec<u8>]) -> Self {
+        assert!(data.len() > 1, "Data must have at least 2 elements");
+        let mut tree = Merkle { tree: vec![] };
+        tree.build(data);
+        tree
+    }
+
+    /// Builds the Merkle Tree using a recursive bottom-up approach.
+    fn build(&mut self, data: &[Vec<u8>]) {
+        // Initialize tree with leaf-level hashes
+        let mut current_level: Vec<Hash> = data.iter().map(|element| hash(element)).collect();
+        let mut tree = current_level.clone();
+
+        // Build tree bottom-up
+        let mut count_leaves = current_level.len();
+        while count_leaves > 1 {
+            let mut level_hashes: Vec<Hash> = vec![];
+            let mut i = 0;
+            while i < count_leaves {
+                if i + 1 < count_leaves {
+                    // Combine adjacent hashes
+                    level_hashes.push(hash_internal_node(&current_level[i], &current_level[i + 1]));
+                } else {
+                    // Clone the last hash to create a pair
+                    level_hashes.push(hash_internal_node(&current_level[i], &current_level[i]));
+                }
+                i += 2;
+            }
```

Comment:
Using a for loop with a range could be more idiomatic
```suggestion
            for i in (0..count_leaves).step_by(2) {
                if i + 1 < count_leaves {
                    // Combine adjacent hashes
                    level_hashes.push(hash_internal_node(&current_level[i], &current_level[i + 1]));
                } else {
                    // Clone the last hash to create a pair
                    level_hashes.push(hash_internal_node(&current_level[i], &current_level[i]));
            }
        }
```

Reply:
Done